WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,216 @@
+
+		public Task AppendEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			IEnumerable<WrappedEvent> wrappedEvents)
+		{
+			Guard.NotNullOrEmpty(nameof(aggregateType), aggregateType);
+			Guard.NotNullOrEmpty(nameof(aggregateId), aggregateId);
+			Guard.NotNull(nameof(wrappedEvents), wrappedEvents);
+			var wrappedEventsList = wrappedEvents.ToList().AsReadOnly();
+			if (wrappedEventsList.Count <= 0)
+			{
+				return Task.CompletedTask;
+			}
+			var firstSequenceId = wrappedEventsList[0].SequenceId;
+			var lastSequenceId = wrappedEventsList[^1].SequenceId;
+			if (firstSequenceId <= 0)
+			{
+				throw new ArgumentException("First event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+			if (lastSequenceId <= 0)
+			{
+				throw new ArgumentException("Last event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+			if (lastSequenceId - firstSequenceId + 1 != wrappedEventsList.Count)
+			{
+				throw new ArgumentException("Event sequence ids are out of whack.", nameof(wrappedEvents));
+			}
+
+			var aggregateEventsBatches = _aggregatesEventsBatches.GetOrAdd(aggregateType, _ => new());
+			var (tailSequenceId, events) = aggregateEventsBatches.GetOrAdd(
+				aggregateId,
+				_ => (0, ImmutableList<WrappedEvent>.Empty));
+
+			if (tailSequenceId + 1 < firstSequenceId)
+			{
+				throw new ArgumentException($"Invalid firstSequenceId (gap in sequence ids) expected: '{tailSequenceId + 1}' given: '{firstSequenceId}'.", nameof(wrappedEvents));
+			}
+
+			// no action
+			Validated();
+
+			var newEvents = events.AddRange(wrappedEventsList);
+
+			// Atomically detect conflict and replace lastSequenceId and lock to ensure strong order in eventsBatches.
+			if (!aggregateEventsBatches.TryUpdate(
+				key: aggregateId,
+				newValue: (lastSequenceId, newEvents),
+				comparisonValue: (firstSequenceId - 1, events)))
+			{
+				Conflicted(); // no action
+				throw new OptimisticConcurrencyException($"Conflict while commiting events. Retry command. aggregate: '{aggregateType}' id: '{aggregateId}'");
+			}
+			Appended(); // no action
+
+			// If it is a first event for given aggregate.
+			if (tailSequenceId == 0)
+			{ // Add index of aggregate id into the dictionary.
+				IndexNewAggregateId(aggregateType, aggregateId);
+			}
+			return Task.CompletedTask;
+		}
+
+		public Task<IReadOnlyList<WrappedEvent>> ListEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			long afterSequenceId = 0,
+			int? limit = null)
+		{
+			if (_aggregatesEventsBatches.TryGetValue(aggregateType, out var aggregateEventsBatches) &&
+				aggregateEventsBatches.TryGetValue(aggregateId, out var value))
+			{
+				var result = value.Events;
+				if (afterSequenceId > 0)
+				{
+					var foundIndex = result.BinarySearch(
+						new WrappedEvent(afterSequenceId),
+						Comparer<WrappedEvent>.Create((a, b) => a.SequenceId.CompareTo(b.SequenceId)));
+					if (foundIndex < 0)
+					{
+						foundIndex = ~foundIndex;
```

Comment:
I don't really know what this line does. Partly because bitwise complement is easier to grasp with unsigned integers. Given that `foundIndex` is `int`, anyone who reads this must consult https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/bitwise-and-shift-operators#bitwise-complement-operator- to find out that it's not described. 

My personal take on this is: It's not worth it. It's better to write it explicitely so that everyone understands it immediately.

Reply:
The [documentation for `BinarySearch`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablelist-1.binarysearch?view=net-6.0#System_Collections_Immutable_ImmutableList_1_BinarySearch__0_System_Collections_Generic_IComparer__0__) states that it returns the bitwise complement of the insertion point if an equal element was not found. Knowing that, reasoning about the equivalence of the `Math.Abs` form requires an understanding of two's complement.

I think it would be better to just add a comment that explains that `~foundIndex` is there because that's what the `BinarySearch` documentation says to do.

Reply:
> I think it would be better to just add a comment that explains that ~foundIndex is there because that's what the BinarySearch documentation says to do.

Sounds good to me.

Reply:
yeah the problem is the interface of BinarySearch I just did what Microsoft asked me to do in documentation. I wanna keep it like that so it is visibly according to documentation